<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Practical MR session using R</title>
    <meta charset="utf-8" />
    <meta name="author" content="Daniele Bottigliengo" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/metropolis.css" rel="stylesheet" />
    <link href="libs/remark-css/metropolis-fonts.css" rel="stylesheet" />
    <script src="libs/htmlwidgets/htmlwidgets.js"></script>
    <script src="libs/viz/viz.js"></script>
    <link href="libs/DiagrammeR-styles/styles.css" rel="stylesheet" />
    <script src="libs/grViz-binding/grViz.js"></script>
  </head>
  <body>
    <textarea id="source">





class: middle, center, title-slide

# Practical MR session using R
&lt;br&gt;

## Daniele Bottigliengo
## Institute for Biomedicine, Eurac Research, Bolzano, Italy
## July 22th, 2021

.pull-left[

&lt;br&gt;
&lt;br&gt; 

&lt;img src="fig/logoEURAC.jpg", width = 30%, heigth = 30%&gt;

]

.pull-right[

&lt;img src="fig/CKDGen_logo_halfsize.png", width = 30%, heigth = 30%&gt;

]
---

# Outline

1. **Introduction**
  - _Two-sample MR framework_
  - _R packages for Two-sample MR_

1. **Data preparation**
  - _Summary GWAS data_
  - _Data harmonization_
  
1. **Performing MR analysis**
  - _`MendelianRandomization` and `TwoSampleMR` functions_
  - _Displaying results_
  
1. **Sensitivity analyses**
  - _Evaluating MR assumptions_
  - _Using MR robust methods_
  
1. **Case study**

---

# Introduction

&lt;br&gt;

- .font140[Two sample MR analysis]

&lt;br&gt;

- .font140[Summary instrument-exposure and instrument-outcome association results]

&lt;br&gt;

- .font140[Summary-level data from GWAS]

&lt;br&gt;

- .font140[Non-overlapping sets of individuals]


---

# R packages

.pull-left[

### [`TwoSampleMR`](https://mrcieu.github.io/TwoSampleMR/index.html)

- GWAS database

- Data harmonization

- Most commonly used MR methods

- Visualization and sensitivity analyses

- Installation from Github:

```r
install.packages("remotes")
remotes::install_github("MRCIEU/TwoSampleMR")
```

]

.pull-right[

### [`MendelianRandomization`](https://cran.r-project.org/web/packages/MendelianRandomization/index.html)

- GWAS database

- Several MR methods are implemented

- Visualization and sensitivity analyses

- Installation from CRAN:

```r
install.packages("MendelianRandomization", dependencies = TRUE)
```

]

---

# Data sources

- .font150[Retrieving summary-level GWAS data:]

  - .font120[[NHGRI-EBI GWAS Catalog](https://www.ebi.ac.uk/gwas/)]
  
  - .font120[[dbGaP CHARGE Consortium GWAS results](https://www.ncbi.nlm.nih.gov/gap/) (*study accession phs000930*)]
  
  - .font120[[Neale lab](http://www.nealelab.is/uk-biobank)]
  
  - .font120[[GeneATLAS](http://geneatlas.roslin.ed.ac.uk/)]
  
  - .font120[[FinnGen cohort](http://r2.finngen.fi/)]

  - .font120[[Phenoscanner](http://www.phenoscanner.medschl.cam.ac.uk/)] `-&gt; MendelianRandomization`

  - .font120[[IEU GWAS database](https://gwas.mrcieu.ac.uk/)] `-&gt; TwoSampleMR`
  
  - GWAS consortium websites, e.g. **CKDGen**, **GIANT**, **MAGIC**,
    **ENIGMA**, **ThyroidOmics**, **COVID-19 HGI**

&lt;br&gt;


---

# IVs selection

&lt;br&gt;

&lt;br&gt;

- .font150[Several strategies for selecting genetic instruments]

&lt;br&gt;


- .font150[Accounting for:]
  - .font130[weak instrument bias]
  - .font130[linkage disequilibrium (LD), i.e. correlations between instruments]
  - .font130[horizontal pleiotropy]
  
---

# IVs selection

1. .font140[Selection of genome-wide significant instruments from GWAS exposure data] `\(\left( p-value &lt; 5\times10^{-8} \right)\)`
  - One can select functional genetic variants, i.e. genome-wide
    significant SNPs in the encoding gene
  - Check `\(F-statistic\)` to evaluate the strength of the instrument

1. .font140[Perform LD clumping to select independent genetic instruments]
  - Default at `\(r^{2} &lt; 0.001\)`
  
1. .font140[Proxies can be used to replace exposure variants that are missing in the outcome dataset]
  - The proxy should be in high LD with the missing genetic instrument
  
1. .font140[Applying Steiger filter to remove IVs for which there might evidence of reverse causation]

---

# Data harmonization

&lt;img src="fig/mr_data_harm.PNG", width = 90%,  align = "center"&gt;

&lt;br&gt;

&lt;https://elifesciences.org/articles/34408&gt;

---

class: middle

# MR analysis

<div id="htmlwidget-3809500ef3e96bf45016" style="width:103%;height:103%;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-3809500ef3e96bf45016">{"x":{"diagram":"digraph{\n                          \n  graph[layout = dot, rankdir = LR]\n  \n  # Define the global shape of the boxes\n  node [shape = rectangle, style = filled, fillcolor = OrangeRed]\n  \n  # Define the nodes\n  primary [label = \"Primary \n analysis\"]\n  iv [label = \"One \n genetic instrument\", style = dashed, shape = plaintext]\n  ivs [label = \"Multiple \n genetic instruments\", style = dashed, shape = plaintext]\n  wald [label = \"Wald estimator\", fillcolor = DarkOrange]\n  ivw [label = \"IVW\", fillcolor = DarkOrange]\n  fe [label = \"IVW \n fixed-effect\", shape = oval]\n  re [label = \"IVW \n random-effect \n (default)\", shape = oval]\n  corr [label = \"IVW \n fixed-effect \n correlated variants\", shape = oval, fillcolor = DarkTurquoise]\n  ce [label = \"Has a causal effect \n been identified?\", style = dashed, shape = plaintext]\n  sens [label = \"Sensitivity analyses\"]\n  \n  \n  # Define the edges\n  primary -> {iv, ivs} iv -> wald ivs -> ivw ivw -> {re, fe}\n  fe -> corr fe -> ce corr -> ce re -> ce ce -> sens\n  \n\n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>

---

# MR analysis

.pull-left[

&lt;img src="fig/scatter_plot.png"&gt;

&lt;small&gt;&lt;small&gt;
&lt;https://mrcieu.github.io/TwoSampleMR/articles/perform_mr.html&gt;
&lt;small&gt;&lt;small&gt;

]

.pull-right[

&lt;img src="fig/forest_plot.png", width = 75%, heigth = 75%&gt;

]

---

# Sensitivity analyses

&lt;br&gt;

- .font140[Evaluation of MR assumptions]

&lt;br&gt;

- .font140[Check for the presence of pleiotropic effects]
  
  + .font120[Heterogeneity of Wald estimates using Q-statistic or ] `\(I^{2}\)`
  + .font120[Intercept of MR-Egger]
  
&lt;br&gt;

- .font140[Leave-one-out analysis to assess the reliance of MR analysis on a particular variant]

---

# Robust MR methods

- Valid causal inferences under weaker assumptions than
the standard IVW methods

- Using multiple methods that make different assumptions
about the nature of the underlying pleiotropy

- Most commonly used robust methods:
  
  + .font90[_MR-Egger_] `\(\rightarrow\)` `TwoSampleMR` and `MendelianRandomization`
  + .font90[_Median based method_] `\(\rightarrow\)` `TwoSampleMR` and `MendelianRandomization`
  + .font90[_Mode based method_] `\(\rightarrow\)` `TwoSampleMR` and `MendelianRandomization`
  + .font90[_MR-PRESSO_]  `\(\rightarrow\)` [`MR-PRESSO`](https://github.com/rondolab/MR-PRESSO) R package
  
  
- A simulation study&lt;sup&gt;1&lt;/sup&gt; has shown that two recent methods have good statistical properties in a wide range of scenarios:
  + .font90[_Contamination mixture_] `\(\rightarrow\)` implemented in `MendelianRandomization` with `mr_conmix` function
  + .font90[_MR-Mix_]  `\(\rightarrow\)` [`MR-Mix`](https://github.com/gqi/MRMix) R package
  
  
.footnote[
.font70[

[1] Slob EAW, Burgess S. **A comparison of robust Mendelian randomization methods using summary data**. _Genetic Epidemiology_. 2020;44:313â€“29.

]
]
  
---

# Other approaches to sensitivity analysis

&lt;br&gt;

- .font140[Using different datasets]

&lt;br&gt;

- .font140[Positive and negative control outcomes]

&lt;br&gt;

- .font140[Subgroup analyses]

&lt;br&gt;

- .font140[Evaluate association with potentially pleiotropic variables]

---

# Case study

&lt;br&gt;

- .font140[**Aim**: to evaluate the causal relationship between kidney function and blood pressure (BP)]

&lt;br&gt;

- .font140[**Exposure**: estimated glomerular filtration rate from serum creatinine (eGFRcr)]

&lt;br&gt;

- .font140[**Outcome**: systolic blood pressure (SBP)]

&lt;br&gt;

---

# Case study


- .font140[Two sets of IVs:]

&lt;br&gt;

  + .font120[256 SNPs associated with eGFRcr at genome-wide threshold] `\(\left( p-value &lt; 5\times10^{-8} \right)\)`

&lt;br&gt;

  + .font120[40 SNPs associated with both eGFRcr and blood urea nitrogen (BUN) at genome-wide threshold] `\(\left( p-value &lt; 5\times10^{-8} \right)\)` `\(\rightarrow\)` prioritizing IVs most likely relevant for 
  kidney function

&lt;br&gt;

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
